packages/sanity/src/core/bundles/hooks/useDocumentVersions.tsx

```diff
@@ -35,92 +29,31 @@ const LISTEN_OPTIONS: ListenOptions = {
-
-  const [error, setError] = useState<Error | undefined>(undefined)
   const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
   const {data: bundles} = useBundles()
   const publishedId = getPublishedId(documentId, documentId.includes('.'))
 
-  const QUERY = `*[sanity::versionOf("${publishedId}")]`
-
-  const initialFetch = useCallback(async () => {
-    if (!client) {
-      return
-    }
-
-    try {
-      const res = await client.fetch(QUERY)
-
-      // build the document bundles
-      const documentBundles = res
-        ?.map((r: BundleDocument) =>
-          bundles?.find((b) => r._version && getBundleSlug(r._id) === b.slug),
-        )
-        .filter((b: BundleDocument) => b !== undefined)
-
-      setState(documentBundles)
-    } catch (err) {
-      setError(err)
-    }
-  }, [QUERY, bundles, client])
+  const QUERY = `*[sanity::versionOf("${publishedId}")]{_id, _version}`
 
-  const listener$ = useMemo(() => {
-    if (!client) return of()
-
-    const events$ = client.observable.listen(QUERY, {}, LISTEN_OPTIONS).pipe(
-      catchError((err) => {
-        setError(err)
-        return of() // return an empty observable on error
-      }),
-    )
-
-    return events$
-  }, [QUERY, client])
-
-  const handleListenerEvent = useCallback(
-    async (event: ListenEvent<Record<string, BundleDocument[]>>) => {
-      if (event.type === 'welcome') {
-        if (!state) {
-          await initialFetch()
-        }
-      }
-
-      if (event.type === 'mutation') {
-        if (event.transition === 'disappear') {
-          const removedDocumentId = getPublishedId(event.documentId)
-          const updatedBundles = state?.filter(
-            (b) => b._id !== removedDocumentId,
-          ) as BundleDocument[]
-          setState(updatedBundles)
-        }
-        const prev = event.result
-        const exists = state?.find((b) => b.slug === getBundleSlug(prev?._id || ''))
-
-        if (!prev) return
-        if (exists) {
-          const updatedBundles = state?.map((b: BundleDocument) =>
-            exists ? prev : b,
-          ) as BundleDocument[]
-          setState(updatedBundles || [])
-        } else {
-          const newBundles = [...(state || []), prev] as BundleDocument[]
-          setState(newBundles)
-        }
-      }
-    },
-    [initialFetch, state],
+  const observable = useMemo(
+    () => listenQuery(client, QUERY) as Observable<Pick<SanityDocument, '_id' | '_version'>[]>,
+    [QUERY, client],
   )
 
-  useEffect(() => {
-    const sub = listener$.subscribe(handleListenerEvent)
+  // This will return the sanity documents that are a version of the published document.
+  const versions = useObservable(observable, null)
+
+  const state = useMemo(
+    () =>
+      versions
+        ? (versions
+            .map((r) => bundles?.find((b) => r._version && getBundleSlug(r._id) === b.slug))
+            .filter(Boolean) as BundleDocument[])
+        : null,
+    [versions, bundles],
+  )
```

Comment:
COULD:
just to improve readability with the find nested in the map, I feel this reads a little more clearly:

```suggestion
  const state = useMemo(() => {
    if (!versions) return null

    const validBundles = versions
      .map(({_id, _version}) => {
        if (!_version) return null
        const bundleSlug = getBundleSlug(_id)
        return bundles?.find((bundle) => bundle.slug === bundleSlug) || null
      })
      .filter(Boolean) as BundleDocument[]

    return validBundles
  }, [versions, bundles])
```

Reply:
Good suggestion, applied, thanks!